File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_record_mappers/BloodGlucoseRecordMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_record_mappers

import androidx.health.connect.client.records.BloodGlucoseRecord
import androidx.health.connect.client.records.MealType
import androidx.health.connect.client.records.metadata.Metadata
import androidx.health.connect.client.units.BloodGlucose
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.BloodGlucoseDto
import com.phamtunglam.health_connector_hc_android.pigeon.BloodGlucoseRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.BloodGlucoseRelationToMealDto
import com.phamtunglam.health_connector_hc_android.pigeon.BloodGlucoseSpecimenSourceDto
import com.phamtunglam.health_connector_hc_android.pigeon.DeviceTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MealTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MetadataDto
import com.phamtunglam.health_connector_hc_android.pigeon.RecordingMethodDto
import io.kotest.matchers.shouldBe
import java.time.Instant
import java.time.ZoneOffset
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("BloodGlucoseRecordMapper")
class BloodGlucoseRecordMapperTest {

    private companion object {
        const val TEST_ID = "blood-glucose-id"
        const val TEST_TIME = 1609459200000L
        const val TEST_GLUCOSE_MMOL = 5.5
        val TEST_ZONE_OFFSET: ZoneOffset = ZoneOffset.ofHours(1)
    }

    @Test
    @DisplayName("GIVEN BloodGlucoseRecord → WHEN toDto → THEN converts correctly")
    fun whenRecordToDto_thenConvertsCorrectly() {
        // Given
        val record = BloodGlucoseRecord(
            level = BloodGlucose.millimolesPerLiter(TEST_GLUCOSE_MMOL),
            specimenSource = BloodGlucoseRecord.SPECIMEN_SOURCE_PLASMA,
            mealType = MealType.MEAL_TYPE_LUNCH,
            relationToMeal = BloodGlucoseRecord.RELATION_TO_MEAL_AFTER_MEAL,
            time = Instant.ofEpochMilli(TEST_TIME),
            zoneOffset = TEST_ZONE_OFFSET,
            metadata = Metadata.manualEntry(),
        )

        // When
        val result = record.toDto()

        // Then
        result.bloodGlucose.millimolesPerLiter shouldBe TEST_GLUCOSE_MMOL
        result.specimenSource shouldBe BloodGlucoseSpecimenSourceDto.PLASMA
        result.mealType shouldBe MealTypeDto.LUNCH
        result.relationToMeal shouldBe BloodGlucoseRelationToMealDto.AFTER_MEAL
        result.time shouldBe TEST_TIME
        result.zoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
    }

    @Test
    @DisplayName("GIVEN BloodGlucoseRecordDto → WHEN toHealthConnect → THEN creates correctly")
    fun whenDtoToHealthConnect_thenCreatesCorrectly() {
        // Given
        val dto = BloodGlucoseRecordDto(
            id = TEST_ID,
            time = TEST_TIME,
            zoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            bloodGlucose = BloodGlucoseDto(
                millimolesPerLiter = TEST_GLUCOSE_MMOL,
            ),
            specimenSource = BloodGlucoseSpecimenSourceDto.PLASMA,
            mealType = MealTypeDto.LUNCH,
            relationToMeal = BloodGlucoseRelationToMealDto.AFTER_MEAL,
            metadata = MetadataDto(
                dataOrigin = "com.example.app",
                recordingMethod = RecordingMethodDto.MANUAL_ENTRY,
                deviceType = DeviceTypeDto.PHONE,
            ),
        )

        // When
        val result = dto.toHealthConnect()

        // Then
        result.level.inMillimolesPerLiter shouldBe TEST_GLUCOSE_MMOL
        result.specimenSource shouldBe BloodGlucoseRecord.SPECIMEN_SOURCE_PLASMA
        result.mealType shouldBe MealType.MEAL_TYPE_LUNCH
        result.relationToMeal shouldBe BloodGlucoseRecord.RELATION_TO_MEAL_AFTER_MEAL
        result.time shouldBe Instant.ofEpochMilli(TEST_TIME)
        result.zoneOffset shouldBe TEST_ZONE_OFFSET
        result.metadata.id shouldBe TEST_ID
    }
}

File: packages/health_connector_hc_android/test/unit_tests/src/mappers/health_record_mappers/blood_glucose/blood_glucose_record_mapper_test.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/blood_glucose/blood_glucose_record_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';
import 'package:test/test.dart';

import '../../../../utils/fake_data.dart';

void main() {
  group(
    'BloodGlucoseRecordMapper',
    () {
      group(
        'BloodGlucoseRecordToDto',
        () {
          test(
            'converts BloodGlucoseRecord to BloodGlucoseRecordDto',
            () {
              // Given
              final record = BloodGlucoseRecord(
                id: HealthRecordId(FakeData.fakeId),
                time: FakeData.fakeTime,
                zoneOffsetSeconds: FakeData.fakeZoneOffsetSeconds,
                metadata: const Metadata(
                  dataOrigin: DataOrigin(FakeData.fakeDataOrigin),
                  recordingMethod: RecordingMethod.activelyRecorded,
                  clientRecordVersion: 1,
                  device: Device(type: DeviceType.phone),
                ),
                bloodGlucose: const BloodGlucose.millimolesPerLiter(5.5),
                relationToMeal: BloodGlucoseRelationToMeal.fasting,
                specimenSource: BloodGlucoseSpecimenSource.capillaryBlood,
                mealType: BloodGlucoseMealType.breakfast,
              );

              // When
              final dto = record.toDto();

              // Then
              expect(dto.id, FakeData.fakeId);
              expect(dto.time, FakeData.fakeTime.millisecondsSinceEpoch);
              expect(dto.zoneOffsetSeconds, FakeData.fakeZoneOffsetSeconds);
              expect(dto.metadata.dataOrigin, FakeData.fakeDataOrigin);
              expect(dto.bloodGlucose.value, 5.5);
              expect(
                dto.bloodGlucose.unit,
                BloodGlucoseUnitDto.millimolesPerLiter,
              );
              expect(
                dto.relationToMeal,
                BloodGlucoseRelationToMealDto.fasting,
              );
              expect(
                dto.specimenSource,
                BloodGlucoseSpecimenSourceDto.capillaryBlood,
              );
              expect(dto.mealType, MealTypeDto.breakfast);
            },
          );
        },
      );

      group(
        'BloodGlucoseRecordDtoToDomain',
        () {
          test(
            'converts BloodGlucoseRecordDto to BloodGlucoseRecord',
            () {
              // Given
              final dto = BloodGlucoseRecordDto(
                id: FakeData.fakeId,
                time: FakeData.fakeLocalTime.millisecondsSinceEpoch,
                zoneOffsetSeconds: FakeData.fakeZoneOffsetSeconds,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.manualEntry,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                bloodGlucose: BloodGlucoseDto(
                  value: 110.0,
                  unit: BloodGlucoseUnitDto.milligramsPerDeciliter,
                ),
                relationToMeal: BloodGlucoseRelationToMealDto.afterMeal,
                specimenSource: BloodGlucoseSpecimenSourceDto.plasma,
                mealType: MealTypeDto.lunch,
              );

              // When
              final record = dto.toDomain();

              // Then
              expect(record.id.value, FakeData.fakeId);
              expect(record.time, FakeData.fakeTime);
              expect(record.zoneOffsetSeconds, FakeData.fakeZoneOffsetSeconds);
              expect(
                record.metadata.dataOrigin.packageName,
                FakeData.fakeDataOrigin,
              );
              expect(record.bloodGlucose.inMilligramsPerDeciliter, 110.0);
              expect(
                record.relationToMeal,
                BloodGlucoseRelationToMeal.afterMeal,
              );
              expect(
                record.specimenSource,
                BloodGlucoseSpecimenSource.plasma,
              );
              expect(record.mealType, BloodGlucoseMealType.lunch);
            },
          );

          test(
            'converts BloodGlucoseRecordDto with null id to '
            'domain with none id',
            () {
              // Given
              final dto = BloodGlucoseRecordDto(
                time: FakeData.fakeLocalTime.millisecondsSinceEpoch,
                zoneOffsetSeconds: 0,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.activelyRecorded,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                bloodGlucose: BloodGlucoseDto(
                  value: 100.0,
                  unit: BloodGlucoseUnitDto.milligramsPerDeciliter,
                ),
              );

              // When
              final record = dto.toDomain();

              // Then
              expect(record.id, HealthRecordId.none);
              expect(
                record.relationToMeal,
                BloodGlucoseRelationToMeal.unknown,
              );
              expect(
                record.specimenSource,
                BloodGlucoseSpecimenSource.unknown,
              );
              expect(record.mealType, BloodGlucoseMealType.unknown);
            },
          );
        },
      );
    },
  );
}
